### ISSUE TITLE
Push failed when using self-managed GitLab instance

### ISSUE THREAD

**Issue by miaoz2001** [python-semantic-release/python-semantic-release#693]
## The problem

Hi, 
I am new to the python semantic release, just tried a bit but I like it.
It works all good with only one issue,
We have a self-managed GitLab instance, so when I push it pushed to the wrong url. As in the log it says

```
    raise GitCommandError(redacted_command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git push https://*****:*****@gitlab.com/xxx/django-xxx.git master
  stderr: 'remote: HTTP Basic: Access denied. The provided password or token is incorrect or your account has 2FA enabled and you must use a personal access token instead of a password. See https://gitlab.com/help/topics/git/troubleshooting_git#error-on-git-fetch-http-basic-access-denied
fatal: Authentication failed for 'https://gitlab.com/xxx/django-xxx.git/''
```

the gitlab we have is something like https://gitlab.example.com/.

## Expected behavior

It should authenticate correctly using the self-managed gitlab url

## Environment

```shell
Python 3.9.18
pip 23.2.1 from /usr/local/lib/python3.9/site-packages/pip (python 3.9)
semantic-release, version 8.0.8
asgiref==3.7.2
astroid==2.15.6
async-timeout==4.0.3
azure-core==1.29.3
azure-storage-blob==12.17.0
black==23.7.0
certifi==2023.7.22
cffi==1.15.1
charset-normalizer==3.2.0
click==8.1.7
codecov==2.1.13
coreapi==2.3.3
coreschema==0.0.4
coverage==7.3.0
cryptography==41.0.3
defusedxml==0.7.1
dill==0.3.7
distlib==0.3.7
dj-database-url==2.1.0
Django==4.2.4
django-allauth==0.55.0
django-allow-cidr==0.7.1
django-cors-headers==4.2.0
django-extensions==3.2.3
django-filter==23.2
django-queryset-csv==1.1.0
django-redis==5.3.0
django-rest-swagger==2.2.0
django-storages==1.13.2
djangorestframework==3.14.0
dotty-dict==1.3.1
et-xmlfile==1.1.0
factory-boy==3.3.0
Faker==19.3.1
filelock==3.12.3
flake8==6.1.0
flake8-colors==0.1.9
flake8-isort==6.0.0
gitdb==4.0.10
GitPython==3.1.32
gunicorn==21.2.0
httpie==3.2.2
idna==3.4
importlib-metadata==6.8.0
importlib-resources==5.7.0
isodate==0.6.1
isort==5.12.0
itypes==1.2.0
Jinja2==3.1.2
lazy-object-proxy==1.9.0
markdown-it-py==3.0.0
markdown2==2.4.10
MarkupSafe==2.1.3
mccabe==0.7.0
mdurl==0.1.2
multidict==6.0.4
mypy-extensions==1.0.0
numpy==1.25.2
oauthlib==3.2.2
openapi-codec==1.3.2
openpyxl==3.1.2
packaging==23.1
pandas==2.0.3
pathspec==0.11.2
pipenv==2023.8.28
platformdirs==3.10.0
psycopg2-binary==2.9.7
py-trello==0.19.0
pycodestyle==2.11.0
pycparser==2.21
pydantic==1.10.12
pydotplus==2.0.2
pyflakes==3.1.0
Pygments==2.16.1
PyJWT==2.8.0
pylint==2.17.5
pylint-django==2.5.3
pylint-plugin-utils==0.8.2
pyOpenSSL==23.2.0
pyparsing==3.1.1
PySocks==1.7.1
python-dateutil==2.8.2
python-gitlab==3.15.0
python-semantic-release==8.0.8
python3-openid==3.2.0
pytz==2023.3
PyYAML==6.0.1
redis==5.0.0
requests==2.31.0
requests-oauthlib==1.3.1
requests-toolbelt==1.0.0
rest-pandas==1.1.0
rich==13.5.2
sentry-sdk==1.30.0
shellingham==1.5.3
simplejson==3.19.1
six==1.16.0
smmap==5.0.0
sqlparse==0.4.4
tomli==2.0.1
tomlkit==0.12.1
typing_extensions==4.7.1
tzdata==2023.3
unicodecsv==0.14.1
uritemplate==4.1.1
urllib3==2.0.4
virtualenv==20.24.3
Werkzeug==2.3.7
whitenoise==6.5.0
wrapt==1.15.0
zipp==3.16.2
```

Python build tool(s) is pipenv

## Configuration

```
[tool.semantic_release]
assets = []
commit_message = "{version}\n\nAutomatically generated by python-semantic-release"
commit_parser = "angular"
logging_use_named_masks = false
major_on_zero = true
tag_format = "v{version}"
version_variables = ["manage.py:__version__"]

[tool.semantic_release.branches.main]
match = "(main|master)"
prerelease_token = "rc"
prerelease = false

[tool.semantic_release.changelog]
template_dir = "templates"
changelog_file = "CHANGELOG.md"
exclude_commit_patterns = []

[tool.semantic_release.changelog.environment]
block_start_string = "{%"
block_end_string = "%}"
variable_start_string = "{{"
variable_end_string = "}}"
comment_start_string = "{#"
comment_end_string = "#}"
trim_blocks = false
lstrip_blocks = false
newline_sequence = "\n"
keep_trailing_newline = false
extensions = []
autoescape = true

[tool.semantic_release.commit_author]
env = "GIT_COMMIT_AUTHOR"
default = "semantic-release <semantic-release>"

[tool.semantic_release.commit_parser_options]
allowed_tags = ["build", "chore", "ci", "docs", "feat", "fix", "perf", "style", "refactor", "test"]
minor_tags = ["feat"]
patch_tags = ["fix", "perf"]

[tool.semantic_release.remote]
name = "origin"
type = "gitlab"
ignore_token_for_push = false

[tool.semantic_release.remote.token]
env = "GL_TOKEN"

[tool.semantic_release.publish]
dist_glob_patterns = ["dist/*"]
upload_to_vcs_release = true


```

## Logs

Please provide debug logs for the command you are using with the `-vv` flag, e.g.

```shell
        
                                                                                                                                                
                    ## Fix                                                                                                                      
                                                                                                                                                
                    * fix: change ddd                                                                               
                    ([`2d0b563`](https://gitlab.com/xxx/django-xxx/-/commit/2d0b563d6d714774ed1531648c1a6e48a0fbacff))',                
                    prerelease=False)                                                                                                           
           DEBUG    [semantic_release.hvcs.gitlab] DEBUG helpers._wrapper: create_release(<semantic_release.hvcs.gitlab.Gitlab     helpers.py:44
                    object at 0x7f3df7973df0>, tag='v0.7.4', release_notes='# v0.7.4 (2023-08-31)                                               
                                                                                                                                                
                    ## Fix                                                                                                                      
                                                                                                                                                
                    * fix: change ddd                                                                                   
                    ([`2d0b563`](https://gitlab.com/xxx/django-xxxx/-/commit/2d0b563d6d714774ed1531648c1a6e48a0fbacff))',                
                    prerelease=False)                                                                                                           
           DEBUG    [urllib3.connectionpool] DEBUG connectionpool._new_conn: Starting new HTTPS connection (1):           connectionpool.py:1048
                    gitlab.com:443                                                                                                              
[03:14:25] DEBUG    [urllib3.connectionpool] DEBUG connectionpool._make_request: https://gitlab.com:443 "GET /api/v4/user  connectionpool.py:546
                    HTTP/1.1" 401 30                                                                                                            
           ERROR    [semantic_release.cli.commands.version] ERROR version.version: 401: 401 Unauthorized                          version.py:545
                    ╭──────────────────────────────────── Traceback (most recent call last) ────────────────────────────────────╮               
                    │ /usr/local/lib/python3.9/site-packages/semantic_release/cli/commands/version.py:539 in version            │               
                    │                                                                                                           │               
                    │   536 │   │   │   release=release,                                                                        │               
                    │   537 │   │   )                                                                                           │               
                    │   538 │   │   try:                                                                                        │               
                    │ ❱ 539 │   │   │   release_id = hvcs_client.create_or_update_release(                                      │               
                    │   540 │   │   │   │   tag=new_version.as_tag(),                                                           │               
                    │   541 │   │   │   │   release_notes=release_notes,                                                        │               
                    │   542 │   │   │   │   prerelease=new_version.is_prerelease,                                               │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/semantic_release/helpers.py:52 in _wrapper                         │               
                    │                                                                                                           │               
                    │    49 │   │   │   )                                                                                       │               
                    │    50 │   │   │                                                                                           │               
                    │    51 │   │   │   # Call function                                                                         │               
                    │ ❱  52 │   │   │   result = func(*args, **kwargs)                                                          │               
                    │    53 │   │   │                                                                                           │               
                    │    54 │   │   │   # Log result                                                                            │               
                    │    55 │   │   │   logger.debug("%s -> %s", func.__qualname__, str(result))                                │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/semantic_release/hvcs/gitlab.py:131 in create_or_update_release    │               
                    │                                                                                                           │               
                    │   128 │   │   self, tag: str, release_notes: str, prerelease: bool = False                                │               
                    │   129 │   ) -> str:                                                                                       │               
                    │   130 │   │   try:                                                                                        │               
                    │ ❱ 131 │   │   │   return self.create_release(                                                             │               
                    │   132 │   │   │   │   tag=tag, release_notes=release_notes, prerelease=prerelease                         │               
                    │   133 │   │   │   )                                                                                       │               
                    │   134 │   │   except gitlab.GitlabCreateError:                                                            │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/semantic_release/helpers.py:52 in _wrapper                         │               
                    │                                                                                                           │               
                    │    49 │   │   │   )                                                                                       │               
                    │    50 │   │   │                                                                                           │               
                    │    51 │   │   │   # Call function                                                                         │               
                    │ ❱  52 │   │   │   result = func(*args, **kwargs)                                                          │               
                    │    53 │   │   │                                                                                           │               
                    │    54 │   │   │   # Log result                                                                            │               
                    │    55 │   │   │   logger.debug("%s -> %s", func.__qualname__, str(result))                                │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/semantic_release/hvcs/gitlab.py:94 in create_release               │               
                    │                                                                                                           │               
                    │    91 │   │   :return: The tag of the release                                                             │               
                    │    92 │   │   """                                                                                         │               
                    │    93 │   │   client = gitlab.Gitlab(self.api_url, private_token=self.token)                              │               
                    │ ❱  94 │   │   client.auth()                                                                               │               
                    │    95 │   │   log.info("Creating release for %s", tag)                                                    │               
                    │    96 │   │   # ref: https://docs.gitlab.com/ee/api/releases/index.html#create-a-release                  │               
                    │    97 │   │   client.projects.get(self.owner + "/" + self.repo_name).releases.create(                     │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/gitlab/client.py:375 in auth                                       │               
                    │                                                                                                           │               
                    │    372 │   │   The `user` attribute will hold a `gitlab.objects.CurrentUser` object on                    │               
                    │    373 │   │   success.                                                                                   │               
                    │    374 │   │   """                                                                                        │               
                    │ ❱  375 │   │   self.user = self._objects.CurrentUserManager(self).get()                                   │               
                    │    376 │   │                                                                                              │               
                    │    377 │   │   if hasattr(self.user, "web_url") and hasattr(self.user, "username"):                       │               
                    │    378 │   │   │   self._check_url(self.user.web_url, path=self.user.username)                            │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/gitlab/v4/objects/users.py:143 in get                              │               
                    │                                                                                                           │               
                    │   140 │   _obj_cls = CurrentUser                                                                          │               
                    │   141 │                                                                                                   │               
                    │   142 │   def get(self, **kwargs: Any) -> CurrentUser:                                                    │               
                    │ ❱ 143 │   │   return cast(CurrentUser, super().get(**kwargs))                                             │               
                    │   144                                                                                                     │               
                    │   145                                                                                                     │               
                    │   146 class User(SaveMixin, ObjectDeleteMixin, RESTObject):                                               │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/gitlab/exceptions.py:336 in wrapped_f                              │               
                    │                                                                                                           │               
                    │   333 │   │   @functools.wraps(f)                                                                         │               
                    │   334 │   │   def wrapped_f(*args: Any, **kwargs: Any) -> Any:                                            │               
                    │   335 │   │   │   try:                                                                                    │               
                    │ ❱ 336 │   │   │   │   return f(*args, **kwargs)                                                           │               
                    │   337 │   │   │   except GitlabHttpError as e:                                                            │               
                    │   338 │   │   │   │   raise error(e.error_message, e.response_code, e.response_body) from e               │               
                    │   339                                                                                                     │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/gitlab/mixins.py:154 in get                                        │               
                    │                                                                                                           │               
                    │   151 │   │   """                                                                                         │               
                    │   152 │   │   if TYPE_CHECKING:                                                                           │               
                    │   153 │   │   │   assert self.path is not None                                                            │               
                    │ ❱ 154 │   │   server_data = self.gitlab.http_get(self.path, **kwargs)                                     │               
                    │   155 │   │   if TYPE_CHECKING:                                                                           │               
                    │   156 │   │   │   assert not isinstance(server_data, requests.Response)                                   │               
                    │   157 │   │   │   assert self._obj_cls is not None                                                        │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/gitlab/client.py:828 in http_get                                   │               
                    │                                                                                                           │               
                    │    825 │   │   │   GitlabParsingError: If the json data could not be parsed                               │               
                    │    826 │   │   """                                                                                        │               
                    │    827 │   │   query_data = query_data or {}                                                              │               
                    │ ❱  828 │   │   result = self.http_request(                                                                │               
                    │    829 │   │   │   "get", path, query_data=query_data, streamed=streamed, **kwargs                        │               
                    │    830 │   │   )                                                                                          │               
                    │    831 │   │   content_type = utils.get_content_type(result.headers.get("Content-Type"))                  │               
                    │                                                                                                           │               
                    │ /usr/local/lib/python3.9/site-packages/gitlab/client.py:788 in http_request                               │               
                    │                                                                                                           │               
                    │    785 │   │   │   │   pass                                                                               │               
                    │    786 │   │   │                                                                                          │               
                    │    787 │   │   │   if result.status_code == 401:                                                          │               
                    │ ❱  788 │   │   │   │   raise gitlab.exceptions.GitlabAuthenticationError(                                 │               
                    │    789 │   │   │   │   │   response_code=result.status_code,                                              │               
                    │    790 │   │   │   │   │   error_message=error_message,                                                   │               
                    │    791 │   │   │   │   │   response_body=result.content,                                                  │               
                    ╰───────────────────────────────────────────────────────────────────────────────────────────────────────────╯               
                    GitlabAuthenticationError: 401: 401 Unauthorized                                                                            
           INFO     [semantic_release.cli.github_actions_output] INFO github_actions_output.write_if_possible: not   github_actions_output.py:70
                    writing GitHub Actions output, as no file specified                                                                         
Usage: semantic-release version [OPTIONS]
Try 'semantic-release version -h' for help.

Error: 401: 401 Unauthorized
```

If using GitHub actions, ensure the `root_options` contains the `-vv` flag
for verbosity, and attach your workflow definition.

## Additional context

_Add any other information that could be useful, such as a link to your project
(if public), links to a failing GitHub Action, or an example commit._

**Comment by miaoz2001** [comment]
well, it is not a bug, I will have to just set the domain in the config file, thanks!

**Comment by rnorris7756** [comment]
I came here through google since I had the same issue.  Since it's not explicitly stated how you solved this, there are a couple of remote options that [aren't documented](https://python-semantic-release.readthedocs.io/en/latest/configuration.html#tool-semantic-release-remote).  The list of options can be found in the code [here](https://github.com/python-semantic-release/python-semantic-release/blob/82bfcd3ba619ce6bda0eb79b903be48d3ac5ff2c/semantic_release/cli/config.py#L110).

In my case, since I was using a `pyproject.toml`, I needed to have the following set:

```toml
[tool.semantic_release.remote]
domain = "gitlab.mydomain.com"
```

Where `gitlab.mydomain.com` is my gitlab's domain name.
